File: src/main/java/com/oracle/ocmjea/gof/behavioral/mediator/ChatMediatorImpl.java
package com.oracle.ocmjea.gof.behavioral.mediator;

import java.util.ArrayList;
import java.util.List;

public class ChatMediatorImpl implements ChatMediator {

	private List<Usuario> usuarios = new ArrayList<Usuario>();
	
	@Override
	public void enviarMensagem(String msg, Usuario usuario) {
		for(Usuario u : usuarios){
            // Quem envia não deve receber a própria msg
            if(u != usuario){
                u.receber(msg);
            }
        }
	}

	@Override
	public void addUsuario(Usuario usuario) {
		usuarios.add(usuario);
	}
}

File: design-patterns/src/main/kotlin/designpatterns/behavioral/mediator/ChatMediatorImpl.kt
package designpatterns.behavioral.mediator

class ChatMediatorImpl : ChatMediator {
    private val users = mutableListOf<User>()

    override fun sendMessage(msg: String, user: User) {
        for (u in users) {
            if (u !== user) {
                u.receive(msg)
            }
        }
    }

    override fun addUser(user: User) {
        users.add(user)
    }
}
